### PR TITLE
Notebooks: Save cell connection name in cell metadata

### PR DESCRIPTION
**Pull request by lucyzhang929**
This PR adds support for saving a cell's connection name in the cell metadata.

### CODE REVIEW

**Review by barbaravaldez — COMMENTED** [review]
I'm not very familiar with the code, so this might be a silly question. Why do we have different interfaces for ICellMetadata?

File: src/sql/workbench/services/notebook/browser/models/cell.ts [hunk 1]
```diff
@@ -804,6 +815,9 @@ export class CellModel extends Disposable implements ICellModel {
 			cellJson.metadata.tags = metadata.tags;
 			cellJson.outputs = this._outputs;
 			cellJson.execution_count = this.executionCount ? this.executionCount : null;
+			if (this._configurationService?.getValue('notebook.saveConnectionName')) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/cell.ts:818]: In general we shouldn't be coding around the services being undefined. In normal runtime that should be expected never to happen and for testing we should be using mock/stub services as appropriate instead.

File: src/sql/workbench/contrib/notebook/test/electron-browser/cell.test.ts [hunk 2]
```diff
@@ -1024,4 +1024,20 @@ suite('Cell Model', function (): void {
 		assert(!isEditMode);
 	});
 
+	test('Should read connection name from notebook metadata and use its corresponding connection profile', async function () {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/test/electron-browser/cell.test.ts:1027]: You don't seem to be checking that it's using the actual connection profile?

File: src/sql/workbench/services/notebook/browser/models/cell.ts [hunk 3]
```diff
@@ -272,6 +275,14 @@ export class CellModel extends Disposable implements ICellModel {
 		return this._options.notebook.language;
 	}
 
+	public get savedConnectionName(): string | undefined {
+		return this._savedConnectionName;
+	}
+
+	public get context(): ConnectionProfile | undefined {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/cell.ts:282]: What's this being added for? It doesn't seem to be used anywhere here.

**Review by Charles-Gagnon — APPROVED** [review]

File: src/sql/azdata.proposed.d.ts [hunk 4]
```diff
@@ -81,6 +81,10 @@ declare module 'azdata' {
 		export interface INotebookMetadata {
 			connection_name?: string;
 		}
+
+		export interface ICellMetadata {
+			connection_name?: string;
```
Comment (Charles-Gagnon) [inline src/sql/azdata.proposed.d.ts:86]: So now we have a cell-level connection name and a notebook-level one? What's the scenario for when each would be used?

**Comment by Charles-Gagnon** [comment]
@barbaravaldez Are you referring to the two in `azdata.d.ts` and `azdata.proposed.d.ts`?

The definitions in those files are merged at runtime. The definition in proposed is just adding in the connection_name property to the ICellMetadata interface.

File: src/sql/workbench/contrib/notebook/test/electron-browser/cell.test.ts (hunk 2 above)
Comment (lucyzhang929) [reply src/sql/workbench/contrib/notebook/test/electron-browser/cell.test.ts:1027]: Ah yes I'll change the description.

File: src/sql/workbench/services/notebook/browser/models/cell.ts (hunk 3 above)
Comment (lucyzhang929) [reply src/sql/workbench/services/notebook/browser/models/cell.ts:282]: This is used for backend multi-connection work but I wanted this PR to just be focused on the metadata. Thanks for catching that, I'll remove it for now.

File: src/sql/azdata.proposed.d.ts (hunk 4 above)
Comment (lucyzhang929) [reply src/sql/azdata.proposed.d.ts:86]: Yep. For notebooks with only one connection, we'll use the notebook-level connection name and for notebooks with multiple connections, we'll use the cell-level connection name.

**Review by barbaravaldez — APPROVED** [review]

**Pull request merged**